src/app/models/evt-polymorphic-models.ts

```diff
@@ -0,0 +1,132 @@
+
+interface ViewerSource {
+    getDataType(key: string, data?: Surface[]): ViewerDataType;
+    getSource(source: ViewerDataType): ViewerDataInput;
+    getTileSource(change: BehaviorSubject<ViewerDataInput>, http?: HttpClient): Observable<OsdTileSource[]>;
+}
+
+type ViewerConstructors<T> = {
+    [K in keyof T]: new () => T[K];
+};
+
+class ViewerController<T extends Record<string, ViewerSource>> {
+    private factories: T;
+
+    constructor(classes: ViewerConstructors<T>) {
+        this.factories = Object.fromEntries(
+            Object.entries(classes).map(
+                ([key, value]) => ([key, new value()]),
+            ),
+        ) as T;
+    }
+    getSource<K extends keyof T>(source: ViewerDataType, type: string | K): ReturnType<T[K]['getSource']> {
+        return this.factories[type].getSource(source) as ReturnType<T[K]['getSource']>;
+    }
+    getTileSource<K extends keyof T>(change: BehaviorSubject<ViewerDataInput>, type: string | K, http?: HttpClient): ReturnType<T[K]['getTileSource']> {
+        return this.factories[type].getTileSource(change, http) as ReturnType<T[K]['getTileSource']>;
+    }
+    getDataType<K extends keyof T>(type: string, data?: Surface[]): ReturnType<T[K]['getDataType']> {
+        return this.factories[type].getDataType(type, data) as ReturnType<T[K]['getDataType']>;
+    }
+}
+
+class ManifestSource {
+
+    getDataType(key: string): ViewerDataType {
+        return { type: key, value: { manifestURL: AppConfig.evtSettings.files.editionImages[key].value } };
+    }
+
+    getSource(source: ViewerDataType): string {
+        return source.value.manifestURL;
+    }
+
+    getTileSource(change: BehaviorSubject<string>, http: HttpClient): Observable<OsdTileSource[]> {
+        return (
+            change
+                .pipe(
+                    filter((url) => !!url),
+                    distinctUntilChanged(),
+                    switchMap((url) => http.get<{ sequences: Partial<Array<{ canvases }>> }>(url)),
+                    map((manifest) => manifest // get the resource fields in the manifest json structure
+                        .sequences.map((seq) => seq.canvases.map((canv) => canv.images).reduce((x, y) => x.concat(y), []))
+                        .reduce((x, y) => x.concat(y), []).map((res) => res.resource)
+                        .map(this.buildTileSource),
+                    ),
+                )
+        );
+    }
+
+    buildTileSource(manifestResource) {
+        return {
+            '@context': manifestResource.service['@context'],
+            '@id': manifestResource.service['@id'],
+            profile: [manifestResource.service['@profile']],
+            protocol: 'http://iiif.io/api/image',
+            height: manifestResource.height,
+            width: manifestResource.width,
+        };
+    }
+}
+
+class XMLSource {
+
+    getDataType(key: string, data: Surface[]): ViewerDataType {
+        const XMLImages: XMLImagesValues[] = [];
+        data.map(s =>
```

Comment:
Since you're not changing the original array, you can use `forEach`. 
Otherwise you can do
```
XMLImages = data.map(s =>
            s[AppConfig.evtSettings.files.editionImages[key].value]
                ? {
                    url: `${AppConfig.evtSettings.files.imagesFolderUrl}/${s.corresp}.jpg`
                } : {
                    width: s[key][0].width,
                    height: s[key][0].height,
                    url: `${AppConfig.evtSettings.files.imagesFolderUrl}/${s[key][0].url}`
                });
```